doc/memos/rdm-draft-buschsieweke-timer-api.md

```diff
@@ -0,0 +1,259 @@
+        - Only the system time (the portable 64 bit time stamps) need to be synchronized.
+- Utilities for convenience
+    - Timer utilities
+        - Timer that send a message upon expiration
+        - Timer that post an event upon expiration (to allow execution in thread context)
+        - Timer that wakes up a thread upon expiration
+    - Timeout utilities
+        - Timeout for receiving a message
+        - Timeout for locking a mutex
+    - System time utilities
+        - Functions to convert between different time formats, e.g. UNIX and GPS timestamps
+
+## Real-Time Requirements
+
+- The implementation must only disable IRQs for short periods of time. More
+  specifically: With *n* software timers used in parallel, the longest period of
+  time with interrupts disabled be at most *O(n)*
+- The implementation must aim for a minimized number of IRQs caused
+- The implementation must aim for keeping its ISRs short; with *n* software
+  timers used in parallel, the ISR should run for at most *O(n)*
+- The implementation should provide means to execute callbacks in thread
+  context
+
+Justification: Real time capable systems require an upper bound for the latency
+a high level timer implementation provides. For most application - if not all -
+it is trivial to get an upper bound for the number of software timers being
+active in parallel. Hence, an *O(f(n))* latency with *n* being the number of
+active software timers makes the real time behavior of a system predictable and
+easy to estimate. In addition, *f(n)* should increase as slowly as possible,
+while not being to strict to rule out reasonable trade-offs e.g. between
+ROM/RAM requirements and latency. An *O(n)* latency is e.g. feasible to
+implement with a sorted linked list and therefore a reasonable lower bar.
+
+```
+
+## Precision Requirements
+
+- Timer callbacks should never fire prematurely (±1 tick of the underlying
+  timer)
+- The accuracy of execution delay should be measured with single digit micro
+  seconds for short delays in absence of other load on the system
+    - It is acceptable if platform specific tweaking is needed to achieve this
+    - It is acceptable if minimum hardware requirements are needed to achieve
+      this
+- Timer callbacks should fire closely after the timer expired
+    - In absence of other load on the system, accuracy in the order of single
+      digit micro seconds should be achievable when given hardware requirements
+      are fulfilled and platform specific tweaking is done
+
+## Performance Requirements
+
+- Efficient use of RAM/ROM
+- Optimized for power management
```

Comment:
What do you think of elaborating this a little bit?

As the discussion above already pointed out, it is required to turn off underlying hardware timers if they aren't needed to gain power savings.

I'd add something like this (feel free to polish or rephrase):

```
    - The timer implementation should disable underlying hardware in periods when the timer is not used by the application. (I.e. no software timer is running and the timer doesn't back a system time.)
```

Or does this already touches design decisions too much, and thus is out of scope of this RDM?

Reply:
I'd say this is a implementation detail. I fully agree that this is the only sensible way to achieve the goal of proper power management - so only implementations doing just that would IMO be considered for adoption.